src/Voron/Data/BTrees/RecentlyFoundTreePages.cs

```diff
@@ -5,145 +5,252 @@
+            Debug.Assert(Descriptors.Length == CacheSize);
+
+            var matchesByGeneration = FindMatchingByGeneration(_currentGeneration);
+            var matchesByPageNumber = FindMatchingByPageNumber(num);
+
+            var matches = matchesByPageNumber & matchesByGeneration;
+
+            int i = -1;
+            while (matches != 0)
             {
-                var page = _cache[i];
-                page?.Dispose();
+                // We will find the next matching item in the cache and advance the pointer to its rightful place.
+                int advance = BitOperations.TrailingZeroCount(matches) + 1;
+                i += advance;
+
+                // We will advance the bitmask to the location. If there are no more, it will be zero and fail
+                // the re-entry check on the while loop. 
+                matches >>= advance;
+
+                // This just reset the cached page instance. 
+                Descriptors[i].Page = null;
             }
-            
-            Array.Clear(_cache, 0, _cacheSize);
         }
 
-        public void Reset(long num)
+        public void Clear()
         {
-            for (int i = 0; i < _cache.Length; i++)
+            _currentGeneration++;
+            if (_currentGeneration == int.MaxValue)
+                _currentGeneration = 1;
+
+            foreach (ref var descriptor in Descriptors)
             {
-                var page = _cache[i];
-                if (page != null && page.Number == num)
-                {
-                    page.Page = null;
-                    return;
-                }
+                descriptor.Page = null;
             }
         }
+
+        public void Add(TreePage page, SliceOptions firstKeyOption, ReadOnlySpan<byte> firstKey, SliceOptions lastKeyOption, ReadOnlySpan<byte> lastKey, ReadOnlySpan<long> cursorPath)
+        {
+            // PERF: The idea behind this check is that IF the page has a bigger cursor path than what we support
+            // on the struct, and this been a performance improvement we are going to reject it outright. We will
+            // not try to accomodate variability in this regard in order to avoid allocations.
+            if (cursorPath.Length > FoundTreePageDescriptor.MaxCursorPath || 
+                firstKey.Length + lastKey.Length > FoundTreePageDescriptor.MaxKeyStorage)
+                return;
+
+            Debug.Assert(_currentGeneration > 0);
+            Debug.Assert(Descriptors.Length == CacheSize);
+
+            var currentPage = page.PageNumber;
+
+            var matchesByGeneration = FindMatchingByGeneration(_currentGeneration);
+            var matchesByPageNumber = FindMatchingByPageNumber(currentPage);
+
+            var match = matchesByPageNumber & matchesByGeneration;
+
+            // We will find the next matching item in the cache or get a new location.
+            var position = match == 0 ? _current % CacheSize : (uint)BitOperations.TrailingZeroCount(match);
+
```

Comment:
If we already have the page in the cache with another generation, why don't we overwrite it?
Can just drop `FindMatchingByGeneration`, no?

Reply:
Then we'll be assured that for every page, there is just one instance (regardless of generation)

Reply:
That's correct, applying the change to retest.